sdk/servicebus/azure-servicebus/azure/servicebus/_common/auto_lock_renewer.py

```diff
@@ -0,0 +1,132 @@
+        return self
+
+    def __exit__(self, *args):
+        self.shutdown()
+
+    def _renewable(self, renewable):
+        if self._shutdown.is_set():
+            return False
+        if hasattr(renewable, 'settled') and renewable.settled:
+            return False
+        if isinstance(renewable, ServiceBusSession) and not renewable._receiver._running:
+            return False
+        if renewable.expired:
+            return False
+        return True
+
+    def _auto_lock_renew(self, renewable, starttime, timeout, on_lock_renew_failure=None):
+        _log.debug("Running lock auto-renew thread for %r seconds", timeout)
+        error = None
+        try:
+            while self._renewable(renewable):
+                if (utc_now() - starttime) >= datetime.timedelta(seconds=timeout):
+                    _log.debug("Reached auto lock renew timeout - letting lock expire.")
+                    raise AutoLockRenewTimeout("Auto-renew period ({} seconds) elapsed.".format(timeout))
+                if (renewable.locked_until_utc - utc_now()) <= datetime.timedelta(seconds=self.renew_period):
+                    _log.debug("%r seconds or less until lock expires - auto renewing.", self.renew_period)
+                    renewable.renew_lock()
+                time.sleep(self.sleep_time)
+        except AutoLockRenewTimeout as e:
+            renewable.auto_renew_error = e
+        except Exception as e:  # pylint: disable=broad-except
+            _log.debug("Failed to auto-renew lock: %r. Closing thread.", e)
+            error = AutoLockRenewFailed(
+                "Failed to auto-renew lock",
+                inner_exception=e)
+            renewable.auto_renew_error = error
+        finally:
+            if on_lock_renew_failure:
+                if self._shutdown.is_set() \
+                        or (hasattr(renewable, 'settled') and renewable.settled) \
+                        or (isinstance(renewable, ServiceBusSession) and not renewable._receiver._running) \
+                        or (not error and not renewable.expired):
+                    # Basically we want to make sure that any of the "intentional" exit cases are not fired on.
+                    # e.g. shutdown, settlement, session closure, renewer timeout (as opposed to renew failure/expiry)
+                    return
+                on_lock_renew_failure(renewable)
+
+    def register(self, renewable, timeout=300, on_lock_renew_failure=None):
+        """Register a renewable entity for automatic lock renewal.
+
+        :param renewable: A locked entity that needs to be renewed.
+        :type renewable: ~azure.servicebus.ReceivedMessage or
+         ~azure.servicebus.Session
+        :param float timeout: A time in seconds that the lock should be maintained for.
+         Default value is 300 (5 minutes).
+        :param Callable[[Union[Session,ReceivedMessage]], None] on_lock_renew_failure: A callback may be specified
```

Comment:
should `on_lock_renew_failure` accept the error as well? -- `on_lock_renew_failure(renewable, error)`.

such that user doesn't need to go through `renewable.auto_renew_error` to see what error happened.

Reply:
I considered that; but given not all of our "failure cases" have an error (e.g. if you simply have your renew_period set to something wacky and the lock expires on its own) so it seemed less well-aligned than checking the inner error.

This is biased by my thinking that the simpler contract (one parameter) is by its nature preferrable, but if it's a negligable difference I could go either way.  @annatisch  any thoughts here as well?  Adam makes a compelling point.

Reply:
Suppose I'm a user and I registered a `on_lock_renew_failure`.
The intuition I get from the callback name is that it should give me some error information about the failure.
However, it just gives me back the message/session object, and I need to find the inner exception by myself which looks a bit weird to me.

Reply:
Fair point.  You've convinced me to take my position on this from "strongly prefer" to "lukewarm."  Added this to my list of items to discuss for arch with Anna.

Reply:
Per arch discussion, settled on the two-parameter callback.